frame/alliance/src/lib.rs

```diff
@@ -0,0 +1,937 @@
+		pub founders: Vec<T::AccountId>,
+		pub fellows: Vec<T::AccountId>,
+		pub allies: Vec<T::AccountId>,
+		pub phantom: PhantomData<(T, I)>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config<I>, I: 'static> Default for GenesisConfig<T, I> {
+		fn default() -> Self {
+			Self {
+				founders: Vec::new(),
+				fellows: Vec::new(),
+				allies: Vec::new(),
+				phantom: Default::default(),
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config<I>, I: 'static> GenesisBuild<T, I> for GenesisConfig<T, I> {
+		fn build(&self) {
+			for m in self
+				.founders
+				.iter()
+				.chain(self.fellows.iter())
+				.chain(self.allies.iter())
+			{
+				assert!(
+					Pallet::<T, I>::has_identity(m).is_ok(),
+					"Member does not set identity!"
+				);
+			}
+
+			if !self.founders.is_empty() {
+				assert!(
+					!Pallet::<T, I>::has_member(MemberRole::Founder),
+					"Founders are already initialized!"
+				);
+				Members::<T, I>::insert(MemberRole::Founder, self.founders.clone());
+			}
+			if !self.fellows.is_empty() {
+				assert!(
+					!Pallet::<T, I>::has_member(MemberRole::Fellow),
+					"Fellows are already initialized!"
+				);
+				Members::<T, I>::insert(MemberRole::Fellow, self.fellows.clone());
+			}
+			if !self.allies.is_empty() {
+				Members::<T, I>::insert(MemberRole::Ally, self.allies.clone())
+			}
+
+			T::InitializeMembers::initialize_members(
+				&[self.founders.as_slice(), self.fellows.as_slice()].concat(),
+			)
+		}
+	}
+
+	/// The IPFS cid of the alliance rule.
+	/// Founders and fellows can propose a new rule, other founders and fellows make a traditional
+	/// super-majority votes, vote to determine if the rules take effect.
+	///
+	/// Any founder has a special one-vote veto right to the rule setting.
+	#[pallet::storage]
+	#[pallet::getter(fn rule)]
+	pub type Rule<T: Config<I>, I: 'static = ()> = StorageValue<_, cid::Cid, OptionQuery>;
+
+	/// The current IPFS cids of the announcements.
+	#[pallet::storage]
+	#[pallet::getter(fn announcements)]
+	pub type Announcements<T: Config<I>, I: 'static = ()> =
+		StorageValue<_, Vec<cid::Cid>, ValueQuery>;
+
+	/// Maps member and their candidate deposit.
+	#[pallet::storage]
+	#[pallet::getter(fn deposit_of)]
+	pub type DepositOf<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, BalanceOf<T, I>, OptionQuery>;
+
+	/// The current set of candidates.
+	/// If the candidacy is approved by a motion, then it will become an ally member.
+	#[pallet::storage]
+	#[pallet::getter(fn candidates)]
+	pub type Candidates<T: Config<I>, I: 'static = ()> =
+		StorageValue<_, Vec<T::AccountId>, ValueQuery>;
+
+	/// Maps member type to alliance members, including founder, fellow and ally.
+	/// Founders and fellows can propose and vote on alliance motions,
+	/// and ally can only wait to be elevated to fellow.
+	#[pallet::storage]
+	#[pallet::getter(fn members)]
+	pub type Members<T: Config<I>, I: 'static = ()> =
```

Comment:
Looks like there is no need to be able to iterate all member in a member role onchain, so why not just `map AccountId => Option<MemberRole>`?

Front-end can iterate all entires of `Members`.

Reply:
Actually we need the count of founders and fellows to calculate the threshold of proposal, which will iterate all member storage onchain．

Maybe we can add another storage `VoteableMemberCount`, or just iterate all members to calculate the count of founders and fellows when calling `propose`. What do you think

Reply:
Ok. In that case this is fine.